### PR TITLE
nrf_cloud: unittest: create a test case for nrf_cloud library

### PR DESCRIPTION
**Pull request by tony-le-24**
Create a unit test suite with 6 basic test cases for nrf_cloud_init() function

See NCSDK-16128

Signed-off-by: Tony Le <tony.le@nordicsemi.no>

### CODE REVIEW

**Review by gopiotr — COMMENTED** [review]
Please add also register into CODEOWNERS file with path to test `/tests/subsys/net/lib/nrf_cloud/` and person/people responsible for this test:
https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS

**Comment by tony-le-24** [comment]
Thank @gopiotr, I will add my name to the CODEOWNERS file

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+void test_init_null_handler()
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	nfsm_set_current_state_and_notify(STATE_IDLE, NULL);
+	zassert_equal(-EINVAL, nrf_cloud_init(&params),
```
[the remark this answers is not in the document]
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:31]: I missed that case, and was overwhelmed with mocking. Thanks for the info. Will add that one later on😅

**Comment by glarsennordic** [comment]
Since this is a brand new test suite, it may be a good idea to switch over to the V2 ZTEST API; V1 is on the way out the door

You will need to enable `CONFIG_ZTEST_NEW_API=y`

Full information is here:

https://docs.zephyrproject.org/latest/develop/test/ztest.html

**Comment by tony-le-24** [comment]
@glarsennordic sounds like fun. I'll read the docs and update it to v2 soon.

**Review by plskeggs — APPROVED** [review]
LGTM

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <ztest.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/nrf_cloud.h>
+#include "nrf_cloud_fsm.h"
+
+void test_init_not_idle(void)
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:12]: I recommend adding prefix comments to your test-case functions which explicitly describe the specific failure case you are testing, here, as with the other test cases

**Review by glarsennordic — CHANGES_REQUESTED** [review]
LGTM -- I am a bit on the fence about using `nsfm_set_current_state_and_notify` directly, but I think with some comments it would be fine.

And there is my previous comment about the new ZTEST API, but I wanted to give you a proper review rather than just that comment.

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
```
[the remark this answers is not in the document]
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:19]: noted, will add comments.

File: tests/subsys/net/lib/nrf_cloud/src/main.c (hunk 2 above)
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:12]: and here as well, thanks

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	*  and implemented inside nrf_cloud.c, so mocking would not be
+	*  a good idea. Here it changes the internal state to not idle
+	*/
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
+	zassert_equal(-EACCES, nrf_cloud_init(&params),
+			  "return should be -EACCES when not idle");
+}
+
+/* Verifiy that nrf_cloud_init function will throw an error
+*  of invalid argument (-EINVAL) with a null parameter
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:39]: > with a null parameter

I would phrase this as

> when given null params

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		printk("Failed to connect to cloud");
+		break;
+	default:
+		break;
+	}
+}
+
+/* Verify that nrf_cloud_init function will be success with
+*  the right internal state and a valid event handler
+*/
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:84]: Nit: Slight rephrasing might work well here:
```suggestion
/* Verify that nrf_cloud_init function will succeed given
*  valid internal state and event handler
*/
```

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	/* The function is an interface (declared in nrf_cloud_fsm.h)
+	*  and implemented inside nrf_cloud.c, so mocking would not be
+	*  a good idea. Here it changes the internal state to not idle
+	*/
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:32]: I don't think you need to talk directly about the decision not to mock the `fsm`,

Maybe phrase this as:

> nfsm_set_current_state_and_notify directly sets the internal state of the nrf_cloud lib, allowing us to easily test various initialization scenarios.

**Review by glarsennordic — APPROVED** [review]
LGTM, minor revision suggestions for comments

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+#include "nrf_cloud_fsm.h"
+
+/* Main test function that declares a test suite
+*  includes all the test cases implemented below
+*/
+ZTEST_SUITE(nrf_cloud_init_test, NULL, NULL, NULL, NULL, NULL);
+
+/* The library can only be initialized when the internal state is
+*  not idle, so the test verifies that nrf_loud_init function will
+*  return with permission denied error (-EACCES) with said condition
```
Comment (jayteemo) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:20]:
```suggestion
*  return with permission denied error (-EACCES) when not idle
```

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#include <net/nrf_cloud.h>
+#include "nrf_cloud_fsm.h"
+
+/* Main test function that declares a test suite
+*  includes all the test cases implemented below
+*/
+ZTEST_SUITE(nrf_cloud_init_test, NULL, NULL, NULL, NULL, NULL);
+
+/* The library can only be initialized when the internal state is
+*  not idle, so the test verifies that nrf_loud_init function will
```
[the remark this answers is not in the document]
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:19]: my bad, will fix now

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+#include <net/nrf_cloud.h>
+#include "nrf_cloud_fsm.h"
+
+/* Main test function that declares a test suite
+*  includes all the test cases implemented below
+*/
+ZTEST_SUITE(nrf_cloud_init_test, NULL, NULL, NULL, NULL, NULL);
+
+/* The library can only be initialized when the internal state is
+*  idle, so the test verifies that nrf_loud_init function will
```
Comment (jayteemo) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:19]:
```suggestion
*  idle, so the test verifies that nrf_cloud_init function will
```

Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:19]: oops I missed that little boo boo there, thanks a ton

**Review by jayteemo — APPROVED** [review]

**Comment by trantanen** [comment]
While the choice of unit test framework doesn't matter for these initial tests, I would generally think of using Unity instead of Ztest  to be able to mock the components that are called.

**Comment by tony-le-24** [comment]
I added some more test cases to this PR to mostly complete the unittest for nrf_cloud_init() function

File: tests/subsys/net/lib/nrf_cloud/prj.conf [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+# MQTT
+CONFIG_NETWORKING=y
+CONFIG_MQTT_LIB=y
+CONFIG_MQTT_LIB_TLS=y
+CONFIG_NRF_CLOUD_MQTT=y
+CONFIG_NRF_MODEM_LIB=y
+
+# FOTA
+CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE=y
+CONFIG_ZCBOR=y
```
Comment (jayteemo) [inline tests/subsys/net/lib/nrf_cloud/prj.conf:30]: you shouldnt need to set this or `CONFIG_MBEDTLS_SHA256_C`

Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/prj.conf:30]: So I should remove that SHA256 config?

Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/prj.conf:30]: when I set `CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPATE=y` and build, CMake throws an error that the other 3 needs to be setup so that `FMFU_FDEV` can be `y`

Comment (jayteemo) [reply tests/subsys/net/lib/nrf_cloud/prj.conf:30]: you need to set
`CONFIG_NRF_MODEM_LIB_SYS_INIT=n`

you should _not_ need to set:
`CONFIG_ZCBOR=y`
and
`CONFIG_MBEDTLS_SHA256_C=y`

File: tests/subsys/net/lib/nrf_cloud/prj.conf [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+CONFIG_ZTEST_NEW_API=y
+
+# Dependencies
+CONFIG_NEWLIB_LIBC=y
+CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
+CONFIG_FPU=y
+CONFIG_CJSON_LIB=y
+
+# Mocking
+CONFIG_UNITY=y
```
Comment (trantanen) [inline tests/subsys/net/lib/nrf_cloud/prj.conf:19]: I guess you should choose either ZTest or Unity. With Unity you get automatically CMock which provides the mocking functionality.
You can discuss this within your team and if you need examples, look at SMS library tests at tests/lib/sms.

Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/prj.conf:14]: I changed to Unity only, thanks

File: tests/subsys/net/lib/nrf_cloud/prj.conf (hunk 10 above)
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/prj.conf]: done, thanks

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		"return should be -EACCES when not idle");
+}
+
+/* Verify that nrf_cloud_init function will throw an error
+ * of invalid argument (-EINVAL) when given null params
+ */
+void test_init_null_params(void)
+{
+	/* Set the internal state to idle */
+	nfsm_set_current_state_and_notify(STATE_IDLE, NULL);
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:45]: Rather than set `STATE_IDLE` directly, I suggest calling `nrf_cloud_uninit` after every test, ensuring it will always be in an un-initialized state.

Then, in this test, _assert_ that `nfsm_get_current_state` returns IDLE

This way you minimize directly interfering with the state machine

I suggest you do this in a tearDown function, which I **_think_** you can create by literally just having a function named `void tearDown(void)` -- better confirm that such a function actually runs if you declare it.

You can do the same for other tests

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	/* nfsm_set_current_state_and_notify directly sets the
+	 * internal state of the nrf_cloud lib, allowing us to
+	 * easily test various initialization scenarios
+	 */
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:34]: I wouldn't use `STATE_TOTAL` since it is not a valid state at all

I recommend calling `nrf_cloud_init` twice instead, since the first init call will get the library into a non-idle state

**Review by glarsennordic — APPROVED** [review]
Looking good, just some minor suggestions regarding `nfsm_set_current_state_and_notify`

File: tests/subsys/net/lib/nrf_cloud/src/main.c (hunk 13 above)
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:34]: I think that `nrf_cloud_init` should not be called twice, since it is under test. There is no guarantee it will not fail when being called the first time. `STATE_IDLE` is not a valid state but it can be good enough for testing.

Comment (glarsennordic) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:34]: > There is no guarantee it will not fail when being called the first time

I think your test case `test_init_success` covers this possibility sufficiently, and I lean towards tests that mess with the internal state as little as possible to make sure that what we are testing is actual behavior

Though if you choose to accept my suggestion, one thing I would also add are some assertions: `nfsm_get_current_state` is `STATE_IDLE` before the test runs, and an assertion that `nfsm_get_current_state` is `STATE_INITIALIZED`  after the first call.

Comment (trantanen) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:34]: I agree with @glarsennordic. If you look at SMS tests (same for location library), everything is basically tested through the public API of the library plus the event handlers underneath some libraries that it's using (`at_monitor_dispatch()`). 

If possible, I'd advise always to avoid using internal functions for testing. They might be needed and required in some cases but the idea should be not to use them if you have a way of avoiding it.

Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:34]: Great, I'll see what I can change to limit these internal calls.

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+	}
+}
+
+/* This function runs before each test */
+void setUp(void)
+{
+	/* Uninit to change the internal state to idle */
+	__wrap_nct_uninit_Ignore();
+	nrf_cloud_uninit();
+}
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:42]: I suggest only do this in `tearDown`

Your `nfsm_get_current_state() == STATE_IDLE` assertions have you covered here

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	/* Expect nct_init to run successfully */
+	__wrap_nct_init_ExpectAndReturn(params.client_id, 0);
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:127]: I suggest putting this assertion before the call to `__wrap_nct_init_ExpectAndReturn`

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+	 * being successful and check if the second one
+	 * will fail since the state will not be idle
+	 */
+	__wrap_nct_init_IgnoreAndReturn(0);
+	nrf_cloud_init(&params);
+
+	int ret = nrf_cloud_init(&params);
+
+	TEST_ASSERT_NOT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should not be idle");
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:76]: Perhaps rephrase this as:
"nrf_cloud lib should not be in the idle state after initialization"

File: tests/subsys/net/lib/nrf_cloud/src/main.c [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+ */
+void test_init_not_idle(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should be idle");
```
Comment (glarsennordic) [inline tests/subsys/net/lib/nrf_cloud/src/main.c:64]: Perhaps rephrase this (and others) as:
"nrf_cloud lib should be in the idle state (not initialized) at the start of test"

**Review by glarsennordic — APPROVED** [review]
Looks great, I very much like the change away from manipulating internal state directly!

Only minor comments on assert placement and phrasing

File: tests/subsys/net/lib/nrf_cloud/src/main.c (hunk 17 above)
Comment (tony-le-24) [reply tests/subsys/net/lib/nrf_cloud/src/main.c:64]: Chaged to "nrf_cloud lib should be in the idle state (uninitialized) at the start of test"

**Review by plskeggs — APPROVED** [review]
LGTM

File: doc/nrf/releases/release-notes-changelog.rst [hunk 18]
```diff
@@ -658,6 +658,7 @@ Libraries for networking
       * :c:func:`nrf_cloud_fota_is_type_enabled` function that determines if the specified FOTA type is enabled by the configuration.
       * :c:func:`nrf_cloud_gnss_msg_json_encode` function that encodes GNSS data (PVT or NMEA) into an nRF Cloud device message.
       * :c:func:`nrf_cloud_fota_pending_job_type_get` function that retreives the FOTA type of a pending FOTA job.
+      * Added unit test for nrf_cloud_init.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:661]:
```suggestion
      * Added unit test for the :c:func:`nrf_cloud_init` function.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 18 above)
Comment (tony-le-24) [reply doc/nrf/releases/release-notes-changelog.rst]: Thank you.

**Review by peknis — APPROVED** [review]
